File: demo-application/demo-sns/backend/app/Http/Controllers/Api/FollowController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\User;
use Illuminate\Http\Request;

class FollowController extends Controller
{
    public function toggle(User $user, Request $request)
    {
        $currentUser = $request->user();

        if ($currentUser->id === $user->id) {
            return response()->json([
                'message' => 'You cannot follow yourself',
            ], 400);
        }

        $follow = $currentUser->following()
            ->where('following_id', $user->id)
            ->first();

        if ($follow) {
            // Unfollow
            $follow->delete();
            $following = false;
        } else {
            // Follow
            $currentUser->following()->create([
                'following_id' => $user->id,
            ]);
            $following = true;
        }

        $user->loadCount(['followers', 'following']);

        return response()->json([
            'following' => $following,
            'followers_count' => $user->followers_count,
            'following_count' => $user->following_count,
        ]);
    }
}

File: demo-application/demo-sns/backend/controllers/follow_controller.go
package controllers

import (
	"demo-sns-backend/database"
	"demo-sns-backend/middleware"
	"demo-sns-backend/models"
	"net/http"

	"github.com/labstack/echo/v4"
	"gorm.io/gorm"
)

type FollowController struct{}

func NewFollowController() *FollowController {
	return &FollowController{}
}

func (fc *FollowController) Toggle(c echo.Context) error {
	userID := c.Param("id")
	currentUser, _ := middleware.GetCurrentUser(c)

	var targetUser models.User
	if err := database.DB.First(&targetUser, userID).Error; err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "User not found"})
	}

	if targetUser.ID == currentUser.ID {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "You cannot follow yourself"})
	}

	var follow models.Follow
	err := database.DB.Where("follower_id = ? AND following_id = ?", currentUser.ID, targetUser.ID).First(&follow).Error

	if err == gorm.ErrRecordNotFound {
		follow = models.Follow{FollowerID: currentUser.ID, FollowingID: targetUser.ID}
		database.DB.Create(&follow)
		return c.JSON(http.StatusOK, map[string]interface{}{"message": "User followed", "following": true})
	} else if err == nil {
		database.DB.Delete(&follow)
		return c.JSON(http.StatusOK, map[string]interface{}{"message": "User unfollowed", "following": false})
	}
	return c.JSON(http.StatusInternalServerError, map[string]string{"error": "Database error"})
}
